### PR TITLE
AI SQL editor

### PR DESCRIPTION
**Pull request by gregnr**
Integrates AI into the SQL editor. Supports:
- Creating new SQL snippets based on a prompt
- Editing existing SQL snippets based on a prompt (using a diff editor)
- Debugging SQL errors using AI
- Generating SQL snippet titles and descriptions using AI

## Staging link
https://supabase-studio-staging-git-feat-ai-sql-editor-supabase.vercel.app/project/_/sql

## Example
image
image
image
image
image
image

### CODE REVIEW

File: studio/components/interfaces/SQLEditor/RenameQueryModal.tsx [hunk 1]
```diff
@@ -1,25 +1,39 @@
...
+
+  const { data: subscription } = useOrgSubscriptionQuery({ orgSlug: slug })
+
+  // Customers on HIPAA plans should not have access to Supabase AI
+  const isHipaaSubscription = subscription?.plan?.id === 'hipaa'
 
   const { id, name, description } = snippet
 
-  const validate = (values: any) => {
+  const [nameInput, setNameInput] = useState(name)
```
Comment (joshenlim) [inline studio/components/interfaces/SQLEditor/RenameQueryModal.tsx:31]: is there a reason why we're using a state here and not the Form's state?

File: studio/components/interfaces/SQLEditor/RenameQueryModal.tsx [hunk 2]
```diff
@@ -56,10 +75,61 @@ const RenameQueryModal = ({ snippet, visible, onCancel, onComplete }: RenameQuer
...
+                {supabaseAIEnabled && !isHipaaSubscription && isAiButtonVisible && (
+                  <Button
+                    type="default"
+                    onClick={async () => {
+                      try {
+                        const { title, description } = await titleSql({
+                          sql: snippet.content.sql,
+                        })
+
+                        setNameInput(title)
```
Comment (joshenlim) [inline studio/components/interfaces/SQLEditor/RenameQueryModal.tsx:95]: there's a helper method from the Form component `resetForm` that we can leverage on, rather than declaring a separate useState

File: studio/components/interfaces/SQLEditor/RenameQueryModal.tsx [hunk 3]
```diff
@@ -56,10 +75,61 @@ const RenameQueryModal = ({ snippet, visible, onCancel, onComplete }: RenameQuer
...
+
+                        // Only update description if it was empty
+                        if (!descriptionInput) {
+                          setDescriptionInput(description)
+                        }
+                      } catch (error: unknown) {
+                        if (isError(error)) {
+                          ui.setNotification({
+                            category: 'error',
+                            message: error.message,
```
Comment (joshenlim) [inline studio/components/interfaces/SQLEditor/RenameQueryModal.tsx:105]:
```suggestion
                            message: `Failed to rename query: ${error.message}`,
```

File: studio/components/interfaces/SQLEditor/RenameQueryModal.tsx (hunk 2 above)
Comment (joshenlim) [reply studio/components/interfaces/SQLEditor/RenameQueryModal.tsx:95]: we can leave this as it is for now since either way we're looking to deprecate the use of this Form component
and use the new one that Jonny has done with shadcn

File: studio/components/interfaces/SQLEditor/UtilityPanel/UtilityActions.tsx [hunk 4]
```diff
@@ -21,7 +27,7 @@ const UtilityActions = ({ id, isExecuting = false, executeQuery }: UtilityAction
       <SizeToggleButton id={id} />
       <Button
         onClick={() => executeQuery()}
-        disabled={isExecuting}
```
Comment (joshenlim) [inline studio/components/interfaces/SQLEditor/UtilityPanel/UtilityActions.tsx:24]: should be disabled as well while executing to prevent multiple clicks?

File: studio/data/ai/sql-debug-mutation.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  return response
+}
+
+type SqlDebugData = Awaited<ReturnType<typeof debugSql>>
+
+export const useSqlDebugMutation = ({
+  onSuccess,
+  ...options
+}: Omit<UseMutationOptions<SqlDebugData, unknown, SqlDebugVariables>, 'mutationFn'> = {}) => {
```
Comment (joshenlim) [inline studio/data/ai/sql-debug-mutation.ts:35]: Use ResponseError instead of unknown for the error type

File: studio/data/ai/sql-edit-mutation.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+  return response
+}
+
+type SqlEditData = Awaited<ReturnType<typeof editSql>>
+
+export const useSqlEditMutation = ({
+  onSuccess,
+  ...options
+}: Omit<UseMutationOptions<SqlEditData, unknown, SqlEditVariables>, 'mutationFn'> = {}) => {
```
Comment (joshenlim) [inline studio/data/ai/sql-edit-mutation.ts:34]: likewise use response error

File: studio/data/ai/sql-debug-mutation.ts (hunk 5 above)
Comment (joshenlim) [reply studio/data/ai/sql-debug-mutation.ts:35]: will resolve in https://github.com/supabase/supabase/pull/16359

File: studio/data/ai/sql-edit-mutation.ts (hunk 6 above)
Comment (joshenlim) [reply studio/data/ai/sql-edit-mutation.ts:34]: will resolve in https://github.com/supabase/supabase/pull/16359

File: studio/components/interfaces/SQLEditor/UtilityPanel/UtilityActions.tsx (hunk 4 above)
Comment (joshenlim) [reply studio/components/interfaces/SQLEditor/UtilityPanel/UtilityActions.tsx:24]: will resolve in will resolve in https://github.com/supabase/supabase/pull/16359

File: studio/pages/api/ai/sql/generate.ts [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+      `,
+    })
+  }
+
+  completionMessages.push({
+    role: 'user',
+    content: prompt,
+  })
+
+  console.log({ completionMessages })
```
Comment (alaister) [inline studio/pages/api/ai/sql/generate.ts:88]: unnecessary console.log?

File: studio/pages/api/ai/sql/generate.ts [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    return res.status(500).json({
+      error: 'There was an unknown error generating the SQL snippet. Please try again.',
+    })
+  }
+
+  const completionResponse: CreateChatCompletionResponse = await response.json()
+
+  console.log(completionResponse)
```
Comment (alaister) [inline studio/pages/api/ai/sql/generate.ts:133]: same here?

File: studio/pages/api/ai/sql/title.ts [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    return res.status(500).json({
+      error: 'There was an unknown error generating the snippet title. Please try again.',
+    })
+  }
+
+  const completionResponse: CreateChatCompletionResponse = await response.json()
+
+  console.log(completionResponse)
```
Comment (alaister) [inline studio/pages/api/ai/sql/title.ts:111]: another

File: studio/pages/api/ai/sql/debug.ts [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+    max_tokens: maxCompletionTokenCount,
+    temperature: 0,
+    function_call: {
+      name: completionFunctions.debugSql.name,
+    },
+    functions: [completionFunctions.debugSql],
+    stream: false,
+  }
+
+  console.log({ sql, completionMessages })
```
Comment (alaister) [inline studio/pages/api/ai/sql/debug.ts:112]: console.log

File: studio/pages/api/ai/sql/debug.ts [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+    }
+
+    return res.status(500).json({
+      error: 'There was an unknown error debugging the SQL snippet. Please try again.',
+    })
+  }
+
+  const completionResponse: CreateChatCompletionResponse = await response.json()
+
+  console.log(completionResponse)
```
Comment (alaister) [inline studio/pages/api/ai/sql/debug.ts:156]: console.log

File: studio/pages/api/ai/sql/debug.ts [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+    console.error(
+      `AI SQL debugging failed: OpenAI response succeeded, but response format was incorrect`
+    )
+
+    return res.status(500).json({
+      error: 'There was an unknown error debugging the SQL snippet. Please try again.',
+    })
+  }
+
+  console.log({ sqlResponseString })
```
Comment (alaister) [inline studio/pages/api/ai/sql/debug.ts:172]: console.log

File: studio/pages/api/ai/sql/edit.ts [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+    max_tokens: maxCompletionTokenCount,
+    temperature: 0,
+    function_call: {
+      name: completionFunctions.editSql.name,
+    },
+    functions: [completionFunctions.editSql],
+    stream: false,
+  }
+
+  console.log({ sql, prompt, completionMessages })
```
Comment (alaister) [inline studio/pages/api/ai/sql/edit.ts:104]: console.log

File: studio/pages/api/ai/sql/edit.ts [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    return res.status(500).json({
+      error: 'There was an unknown error editing the SQL snippet. Please try again.',
+    })
+  }
+
+  const completionResponse: CreateChatCompletionResponse = await response.json()
+
+  console.log(completionResponse)
```
Comment (alaister) [inline studio/pages/api/ai/sql/edit.ts:147]: console.log

File: studio/pages/api/ai/sql/edit.ts [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+    console.error(
+      `AI SQL editing failed: OpenAI response succeeded, but response format was incorrect`
+    )
+
+    return res.status(500).json({
+      error: 'There was an unknown error editing the SQL snippet. Please try again.',
+    })
+  }
+
+  console.log({ sqlResponseString })
```
Comment (alaister) [inline studio/pages/api/ai/sql/edit.ts:163]: console.log

File: studio/data/ai/keys.ts [hunk 16]
```diff
@@ -0,0 +1 @@
+export const aiKeys = {}
```
Comment (alaister) [inline studio/data/ai/keys.ts:1]: is this file needed if there aren't any keys?

**Review by alaister — APPROVED** [review]
LGTM! A few small nits but nothing blocking

File: studio/components/interfaces/SQLEditor/RenameQueryModal.tsx [hunk 17]
```diff
@@ -16,10 +16,15 @@ export interface RenameQueryModalProps {
...
   const { ui } = useStore()
-  const { ref } = useParams()
+  const { ref, slug } = useParams()
   const snap = useSqlEditorStateSnapshot()
   const supabaseAIEnabled = useFlag('sqlEditorSupabaseAI')
 
+  const { data: subscription } = useOrgSubscriptionQuery({ orgSlug: slug })
+
+  // Customers on HIPAA plans should not have access to Supabase AI
+  const isHipaaSubscription = subscription?.plan?.id === 'hipaa'
```
Comment (kevcodez) [inline studio/components/interfaces/SQLEditor/RenameQueryModal.tsx:27]: HIPAA is an add-on, not an individual plan, here's a sample response, stripped down to the important bits from the subscription (same response for org and project endpoints)

```json
{
    "plan": {
        "id": "pro",
        "name": "Pro",
        "price": 25
    },
    "addons": [
        {
            "name": "HIPAA",
            "supabase_prod_id": "addon_security_hipaa",
            "price": 299
        }
    ]
}
```

So the proper way to check it would be 

`subscriptions.addons.some(it => it.supabase_prod_id === 'addon_security_hipaa')`
